torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py

```diff
@@ -14,69 +14,116 @@ def _quantize_array(
+    out_ref: jax.Array,  # (batch_block_size, out_block_size)
+    acc_scratch: jax.Array,  # (batch_block_size, out_block_size)
+    q_x_scratch: jax.Array,  # (batch_block_size, in_block_size)
+    x_scale_scratch: jax.Array,  # (batch_block_size, 1)
     *,
-    quantize_activation,
-    batch_block_size,
-    out_block_size,
-    in_block_size,
+    quantize_activation: bool,
+    save_acc: bool,
+    save_quant: bool,
+    batch_block_size: int,
+    out_block_size: int,
+    in_block_size: int,
 ):
   bs_idx, out_idx, in_idx = pl.program_id(0), pl.program_id(1), pl.program_id(2)
-  nsteps = pl.num_programs(2)
+  n_bs, n_out, n_in = pl.num_programs(0), pl.num_programs(1), pl.num_programs(2)
   x_ref_dtype = x_ref.dtype
   assert x_ref.shape == (batch_block_size,
                          in_block_size), "x_ref shape is not correct"
   assert w_ref.shape == (out_block_size,
                          in_block_size), "w_ref shape is not correct"
   assert scalar_ref.shape == (1,
                               out_block_size), "scalar_ref shape is not correct"
-  assert x_abs_max_val.shape == (
+  assert x_abs_max_ref.shape == (
       1, batch_block_size), "x_max_val shape is not correct"
   assert out_ref.shape == (batch_block_size,
                            out_block_size), "out_ref shape is not correct"
-  assert acc_ref.shape == (batch_block_size,
-                           out_block_size), "acc_ref shape is not correct"
-
-  @pl.when(in_idx == 0)
-  def _():
-    acc_ref[...] = jnp.zeros_like(acc_ref)
-
-  if quantize_activation:
-    x, x_scale = _quantize_array(x_ref[...], x_abs_max_val[...])
-    acc_ref[...] += jax.lax.dot_general(
-        x,
-        w_ref[...],
-        (((1,), (1,)), ((), ())),
-        preferred_element_type=jnp.int32,
-    )
+
+  if save_quant:
+    assert quantize_activation
+    assert q_x_scratch is not None
+    assert x_scale_scratch is not None
+    quant = out_idx == 0
   else:
-    acc_ref[...] += jax.lax.dot_general(
-        x_ref[...],
-        w_ref[...],
-        (((1,), (1,)), ((), ())),
-    )
-
-  @pl.when(in_idx == nsteps - 1)
-  def _():
-    acc = acc_ref[...]
-    scalar = scalar_ref[...]
-    acc *= scalar
+    assert q_x_scratch is None
+    assert x_scale_scratch is None
+    quant = quantize_activation
+
+  if save_acc:
+    assert acc_scratch is not None
+    is_first_step = in_idx == 0
+    is_last_step = in_idx == n_in - 1
+  else:
+    assert acc_scratch is None
+    is_first_step = True
+    is_last_step = True
+
+  def matmul_body(quant, is_first_step, is_last_step):
     if quantize_activation:
-      acc *= x_scale
-    out_ref[...] = acc.astype(x_ref_dtype)
+      if quant:
+        q_x_tmp, x_scale_tmp = _quantize_array(x_ref[...], x_abs_max_ref[...])
+        if save_quant:
+          q_x_scratch[...] = q_x_tmp
+          x_scale_scratch[...] = x_scale_tmp
+      else:
+        assert save_quant
+        q_x_tmp = q_x_scratch[...]
```

Comment:
does it still have the issue? The q_x_scratch has shape `(batch_block_size, in_block_size)`. So if x has shape [3, 2] and in_block_size=1, then your q_x_scratch can only store one block, right?

Reply:
Yes, as mentioned in above comment, I've got around this problem by only allowing `save_quant=true` when `n_in == 1`. So in your example, `save_quant` will be set to `false`.